Add multi-access

SparklingComet opened the pull request:
*Will add a description later*

SparklingComet committed:
Add ShopType enum

SparklingComet committed:
Add #getShopOwners & #getShopMembers in Utils, based on chest inventory names

SparklingComet committed:
Add UNKNOWN constant to ShopType & add #header() method

SparklingComet committed:
Fix #getShopOwners & #getShopMembers

SparklingComet committed:
Format AdminEventListener & move to Utils#getShopOwners() & Utils#getShopMembers

SparklingComet committed:
Create Utils#setName(InventoryHolder) utility method

SparklingComet committed:
Update BiShopCreateEventListener & BiTradeEventListener

SparklingComet committed:
Update IShopCreateEventListener & ITradeEventListener

SparklingComet committed:
Remove //copy comment from Executor.java

SparklingComet committed:
Update ShopCreateEventListener & TradeEventListener

SparklingComet committed:
Reformat code

SparklingComet committed:
Update /tradeshop break

SparklingComet committed:
Add missing @Override annotation in Executor

KillerOfPie commented on src/main/java/org/shanerx/tradeshop/Utils.java:
```diff
@@ -21,13 +21,13 @@
-import org.bukkit.Bukkit;
-import org.bukkit.ChatColor;
-import org.bukkit.Material;
+import org.bukkit.*;
```
If we are importing all of bukkit shouldn't we get rid of the rest of the imports?

SparklingComet replied:
That was done by my IDE, didn't notice it. Wildcard imports are a *really* bad practice so I will fix this immediately.

KillerOfPie commented on src/main/java/org/shanerx/tradeshop/admin/AdminEventListener.java:
```diff
@@ -40,89 +42,89 @@ public AdminEventListener(TradeShop instance) {
+			}
 			event.setCancelled(true);
 			player.sendMessage(colorize(getPrefix() + plugin.getMessages().getString("no-ts-destroy")));
-			return;
```
Why did we lose this return?

SparklingComet replied:
The method's return type is void and it was the last statement of the if block. After that there is only an else block, so no need for it.

KillerOfPie replied:
k

KillerOfPie commented on src/main/java/org/shanerx/tradeshop/admin/AdminEventListener.java:
```diff
@@ -40,89 +42,89 @@ public AdminEventListener(TradeShop instance) {
 	
-	@EventHandler
+	@EventHandler(priority = EventPriority.HIGH)
 	public void onChestOpen(PlayerInteractEvent e) {
```
I want to play with this a bit to make it also show the name of the inventory as "Shop" or "Chest" when they open it

SparklingComet replied:
What exactly do you mean by that?

KillerOfPie commented on src/main/java/org/shanerx/tradeshop/admin/AdminEventListener.java:
```diff
@@ -40,89 +42,89 @@ public AdminEventListener(TradeShop instance) {
-		if (s.getLine(3).equalsIgnoreCase(e.getPlayer().getName())) {
+			
+		} else if (getShopOwners(block).contains(Bukkit.getOfflinePlayer(e.getPlayer().getUniqueId()))
+				|| getShopMembers(block).contains(Bukkit.getOfflinePlayer(e.getPlayer().getUniqueId()))) {
```
We should either have a method that  does both, or have members return all members and owners?

SparklingComet replied:
Will add a method that returns all members/owners

KillerOfPie commented on src/main/java/org/shanerx/tradeshop/bitrade/BiShopCreateEventListener.java:
```diff
@@ -44,51 +45,50 @@ public BiShopCreateEventListener(TradeShop instance) {
-	@EventHandler
+	@EventHandler(priority = EventPriority.HIGH)
 	public void onSignChange(SignChangeEvent event) throws InterruptedException {
-		//	BlockState state = event.getBlock().getState();
```
This can  go

SparklingComet replied:
I thought so.

KillerOfPie commented on src/main/java/org/shanerx/tradeshop/bitrade/BiShopCreateEventListener.java:
```diff
@@ -44,51 +45,50 @@ public BiShopCreateEventListener(TradeShop instance) {
 		Sign s = (Sign) event.getBlock().getState();
-		if (!(event.getLine(0).equalsIgnoreCase("[BiTrade]"))) {
+		
+		if (!event.getLine(0).equalsIgnoreCase("[BiTrade]")) {
```
do we want a method for checking these as well, or should we make the isTradeShop methods ignore case so they find these too?

SparklingComet replied:
When I initially wrote the plugin (Idk how it looks like now) it automatically changed the case to [Trade]/[iTrade], so #equalsIgnoreCase was superficial. I am not sure how it works now.

KillerOfPie replied:
thats for the creation signs(so if a player doesnt cap them) its like that in all the creation files, but only for the first line

KillerOfPie commented on src/main/java/org/shanerx/tradeshop/bitrade/BiShopCreateEventListener.java:
```diff
@@ -44,51 +45,50 @@ public BiShopCreateEventListener(TradeShop instance) {
+		if (!event.getLine(0).equalsIgnoreCase("[BiTrade]")) {
+			return;
+			
+		} else if (!player.hasPermission(getCreatePerm())) {
```
should be getCreateBiPerm

SparklingComet replied:
Uhm... I don't recall changing it... Good catch anyways, will fix.

KillerOfPie commented on src/main/java/org/shanerx/tradeshop/bitrade/BiShopCreateEventListener.java:
```diff
@@ -44,51 +45,50 @@ public BiShopCreateEventListener(TradeShop instance) {
-        
-		boolean signIsValid = true; // If this is true, the information on the sign is valid!
+		
+		if (!plugin.getAllowedInventories().contains(chest.getType())) {
```
move this into the try and have it throw an exception so we dont have to do the setlines and no chest message twice

SparklingComet replied:
Will do.

SparklingComet replied:
maybe we should create a method in utils to set the lines... I'll add it to the TODO.

KillerOfPie replied:
we can, would have to pass it the sign the trade type tho

SparklingComet commented on src/main/java/org/shanerx/tradeshop/bitrade/BiShopCreateEventListener.java:
```diff
@@ -44,51 +45,50 @@ public BiShopCreateEventListener(TradeShop instance) {
-	@EventHandler
+	@EventHandler(priority = EventPriority.HIGH)
 	public void onSignChange(SignChangeEvent event) throws InterruptedException {
-		//	BlockState state = event.getBlock().getState();
```
Good catch, will remove.

SparklingComet committed:
Fix imports

SparklingComet committed:
Make Utils#getShopMembers return owners too

SparklingComet committed:
Fix BiTrade create permission

SparklingComet committed:
Add backwards compatibility (adds the name on the last line to the owners list)

SparklingComet committed:
Add getShopOwners call inside BiTradeEventListener

SparklingComet committed:
Remove getShopOwners from AdminEventListener now that getShopMembers returns owners too

SparklingComet committed:
Add addOwner and removeOwner to Utils

SparklingComet committed:
Add KillerOfPie to the authors in the plugin.yml file

SparklingComet committed:
Add some Javadoc to Utils - not yet complete

SparklingComet committed:
Fix the name on the sign not being added to #getShopMembers()

KillerOfPie commented on src/main/java/org/shanerx/tradeshop/Utils.java:
```diff
@@ -46,219 +55,381 @@
+	}
+	
+	public List<OfflinePlayer> getShopOwners(Sign s) {
+		Chest c = (Chest) findShopChest(s.getBlock());
```
Needs to use getState()

SparklingComet replied:
Fixed in [this commit](https://github.com/SparklingComet/TradeShop/pull/19/commits/b4fd175bbadc031d7420a5ae2eeb638209826fc9)

KillerOfPie commented on src/main/java/org/shanerx/tradeshop/Utils.java:
```diff
@@ -46,219 +55,381 @@
+	}
+	
+	public List<OfflinePlayer> getShopMembers(Sign s) {
+		Chest c = (Chest) findShopChest(s.getBlock());
```
Needs to use getState()

SparklingComet replied:
Fixed in [this commit](https://github.com/SparklingComet/TradeShop/pull/19/commits/b4fd175bbadc031d7420a5ae2eeb638209826fc9)

KillerOfPie reviewed (changes requested):
These need to get fixed before it works, also can you add the commands for adding owners and members. the break command is a good base to start with.

SparklingComet committed:
Update pomx.ml to 1.12

SparklingComet committed:
Fix casting

SparklingComet committed:
Add add/removeOwner

SparklingComet commented:
I have added the addOwner and removeOwner methods in the last commit, but if someone tries to add someone who is already a member the target will be entered twice. There won't be any undesired effects, but it will take storage space, which is limited.
This will have to be fixed.

SparklingComet committed:
Add /ts addowner/removeowner/addmember/removemember [target]

SparklingComet committed:
Update /ts help message

SparklingComet committed:
Fix /ts arguments check

SparklingComet committed:
Finish Javadoc for Utils

KillerOfPie reviewed (approved):
Looks good, gonna test everything real quick and see how many names we can put in

KillerOfPie committed:
Fixes

SparklingComet committed:
Fix bugs with add/removeOwner/Member

SparklingComet committed:
Fast-forward / Fix merge conflicts from origin

KillerOfPie committed:
Removed bukkit wildcard

KillerOfPie reviewed (dismissed):
Looks good i'm tired AF tho, have we added a check for total users on the chest yet? 
and also a check for if the user is already on the chest when we add them?
also players can be added that don't actually exist, are we going to check for that or just leave it to the players, there are no errors if a name is incorrect.

SparklingComet commented:
Writing from my phone, so I cannot make any changes right now.

> Looks good i'm tired AF tho, have we added a check for total users on the chest yet?
and also a check for if the user is already on the chest when we add them?
also players can be added that don't actually exist, are we going to check for that or just leave it to the players, there are no errors if a name is incorrect.

getShopMembers() returns all members, regardless of whether they're owners or not. You can distinguish owners from members by checking whether getShopOwners also contains that OfflinePlayer.

There is no check like that. I believe it was added to the todo list, but it may have been accidentally moved to 'Completed' by me. We should definitely add it before merging, if you can do that (I cannot), that would be excellent.

Lastly, that can stay as it is, at least in my opinion. I also planned a /ts who commad that shows owners an members, so that players can double check who they added anyway.


Finally, I think there are other todo entries under 'Planned', but I may be wrong. Since I won't be home for a while,
I can only review changes but cannot actually commit, so, if you feel like adding it, I hardly have any complaints. Don't feel obligated to, though.

KillerOfPie commented:
getShopMembers only returns members now, if you want everything I added a getShopUsers method,
and ill start working on the stuff but if you can get 1.5.2 out it would be nice

SparklingComet commented on src/main/java/org/shanerx/tradeshop/Utils.java:
```diff
@@ -46,219 +51,575 @@
+        Inventory inv = ((InventoryHolder) b.getState()).getInventory();
+        String names = inv.getName();
+        for (String m : names.split(";")) {
+            if (m.startsWith("m:")) {
```
I am sure this used to be different. The if check should be removed before merging. Another method returning both owners and members is useless.

SparklingComet commented:
> getShopMembers only returns members now, if you want everything I added a getShopUsers method,
and ill start working on the stuff but if you can get 1.5.2 out it would be nice

I agree about the 1.5 part.

KillerOfPie committed:
Fixes
- Fixes

KillerOfPie committed:
Fixes
- fixed imports

KillerOfPie commented:
had to remake the 1.6 branch to catch it up to master

KillerOfPie committed:
Merge branch '1.6' into multi-access

KillerOfPie committed:
Fixes
- code reformat
- fixes recommended by code analysis(when they made sense)
- fixes for issues caused by 1.5.2

KillerOfPie committed:
Fixes
- Stopped players from being able to trade with their own shops
- Fixed Who command
- Fixed adding/removing owners/members
- Re-Fixed ITrade shops
- Stopped Enable on versions below 1.9
- Changed max-break-distance to max-edit distance
- Changed default out of range message
- removed extra ifs
- fixed overflow cause by getting owners/members on old chests
- shortened getShopUsers

KillerOfPie commented:
This is working at this point there is a problem that must be addressed before it can be released though.
If a player that is not a owner on the shop puts a sign next to the chest it will add them to the chest(or make it theirs)

SparklingComet commented:
Sounds nasty, but should be easy and quick to fix. Just check whether getShopSign returns a valid sign and, if yes, cancel the event with a kind message.
Is it safe to assume that the new patches are working?

SparklingComet commented:
> Looks good, gonna test everything real quick and see how many names we can put in

Anything new about that? We need to implement a limit for that.

KillerOfPie commented:
@SparklingComet already in max default users is 5

KillerOfPie commented:
@SparklingComet problem needs to get fixed for double chests too

SparklingComet commented:
I'll add a short checklist to track stuff to change/fic:
- [ ] Stop non-owners to place (trade)signs on shop chests
- [ ] Locking of doublechests needs to be fixed (?) (Found it in the todo list)

SparklingComet commented:
I'm merging this into ``1.6`` because this is going beyond this PR's scope.

SparklingComet commented:
All fixes need to be pushed to ``1.6-fixes``.

KillerOfPie commented:
@SparklingComet changes introduced in 1.5.2 were already merged and all errors from the were fixed

SparklingComet commented:
@KillerOfPie excellent, I assume it had been tested and everything is working.

I am referencing #21 as it contains fixes for this PR.

KillerOfPie commented:
Si
